ink-wrapper-types/src/lib.rs

```diff
@@ -16,7 +17,42 @@ pub trait SignedConnection<TxInfo, E>: Sync {
     async fn exec(&self, account_id: AccountId, data: Vec<u8>) -> Result<TxInfo, E>;
 }
 
+/// Contracts will use this trait for reading data from the chain - non-mutating methods and fetching events.
 #[async_trait]
-pub trait Connection<E>: Sync {
+pub trait Connection<TxInfo, E>: Sync {
     async fn read<T: scale::Decode>(&self, account_id: AccountId, data: Vec<u8>) -> Result<T, E>;
+
+    async fn get_contract_events(&self, tx_info: TxInfo) -> Result<ContractEvents, E>;
+}
+
+/// Represents a raw event emitted by a contract.
+pub struct ContractEvent {
+    /// The account id of the contract that emitted the event.
+    pub account_id: AccountId,
+    /// The unparsed data of the event.
+    pub data: Vec<u8>,
+}
+
+/// Represents a collection of events emitted by contracts in a single transaction.
+pub struct ContractEvents {
+    pub events: Vec<ContractEvent>,
+}
+
+/// A trait that allows to decode events emitted by a specific contract.
+pub trait EventSource: Copy + Into<AccountId> {
+    /// The type to decode the emitted events into.
+    type Event: scale::Decode;
+}
+
+impl ContractEvents {
+    /// Returns a events emitted by a specific contract.
+    pub fn for_contract<C: EventSource>(&self, contract: C) -> Vec<C::Event> {
+        use scale::Decode as _;
+
+        self.events
+            .iter()
+            .filter(|e| e.account_id == contract.into())
+            .filter_map(|e| C::Event::decode(&mut e.data.as_slice()).ok())
```

Comment:
This will silently "swallow" events that couldn't be decoded with the provided decoder -- isn't it the case that we should be able to decode all of them? i.e. their source is contract's `Instance` (as generated by the `ink-wrapper`) and their type is `Instance::Event` (as generated by the `ink-wrapper`). Under what circumstances do you think it's safe to ignore the failed ones? Honest question :)

Reply:
Hm, right, that's not what I intended. I wanted to ignore events that couldn't be decoded as `ContractEmitted`, but that's obviously not what's happening here. I'm on the fence - do you think we should panic or return a `Result<Vec<C::Event>>` here?

Reply:
I think the expectation is that the generated code will be able to decode _all_ events that for this particular instance of contract (described with the used `metadata.json`). In which case we should error and inform the user. We have two options, as you note, to fail -- `panic!` or `Result<_>`:
1. On one hand, not being able to decode an event is breaking the assumption and hence is unexpected outcome of running the code - should `panic!`.
2. On the other hand, panicking here will likely mean that the caller's flow (backend) panics and exits as well (the panic won't be caught) which is not ideal if the txn submission was part of something bigger.

My personal preference (but it's a soft one and I can be convinced to change my mind) is to panic here and print:
* the `TxInfo` data (like txn hash, block hash, etc.) which can be used to uniquely identify it
* event for which the decoding failed

After all, `ink-wrapper` is still in beta phase so it should be fine. Maybe this will improve our feedback loops as well - i.e. potential users submitting reports (:cough: complaining :cough:) more eagerly if the app crashes, rather than logs errors (which is not even happening as of today due to a lock of actual logging).